### DISCUSSION TITLE
SML Pulse Counter in polling mode

### DISCUSSION THREAD

**Discussion by sensor-freak** [arendst/Tasmota#17613] [Q&A]
Hi,

I recently did some experiments with the SML Pulse Counter feature. I use it to monitor a water meter (with a resolution of 50 pulses per liter). I configure the counter in interrupt mode, and this works quite well. Being curious about the polling mode I made also some experiments comparing the results of poll mode and interrupt mode. The results were not satisfying...

The current SML code for polling counters (in branch development) requires 6 consecutive readings of low after a trailing edge in order to identify a state change. Since the minimum polling interval seems to be something as low as 50ms, this causes a low phase of the input signal of about 300ms in order to be detected as valid state change. This might be much to slow, e.g. for water counters that generate 100 or 1000 (or even 10000) pulses per cubic meter.

This behavior also does not match exactly the [documentation](https://tasmota.github.io/docs/Smart-Meter-Interface/#meter-definition), at least as I understand it. The documentation states that a positive value in the `<parameter>` for the Meter Definition defines the poll interval, but it does not give a cue that you need 6 times the poll interval to detect a stable state change.

Does anybody know what were the reasons for that kind of implementation? And is there anybody out there depending on that behavior?

Kind regards

PS: The code I'm referring to is in file [xsns_53_sml.ino](https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xsns_sensor/xsns_53_sml.ino) in method [SML_Counter_Poll](https://github.com/arendst/Tasmota/blob/8275cb776e4f7f7731d5f9058ba85b6b01256101/tasmota/tasmota_xsns_sensor/xsns_53_sml.ino#L3099), see line [3113ff](https://github.com/arendst/Tasmota/blob/8275cb776e4f7f7731d5f9058ba85b6b01256101/tasmota/tasmota_xsns_sensor/xsns_53_sml.ino#L3113)

**Comment by gemu2015** [comment]
why do you assume minimum interval is 50ms ? it can be as low as 1.
a debounce function is simply a low pass and that is what the code does. poll rate is simply sampling rate.
however since Tasmota timing is not precise enough counter polling is not really recommended and is actually deprecated.
use IRQ mode instead.
i will correct the docs in that way.

**Reply by sensor-freak** [reply]
Deduced it by observation (added some log messages). But obviously the minimum time of 50ms was false observation.

I presume the `counter<x>` and `countern<x>` feature (selectable in the module configuration) are also a polling counters, and are also not feasible for fast signals (~200Hz max). So I will stay with SMI counters in IRQ mode.
